File: app/android/app/src/main/kotlin/com/rivendell/app/AnkiGateway.kt
package com.rivendell.app

import android.content.Context
import android.net.Uri
import android.os.Build
import androidx.core.content.ContextCompat
import com.ichi2.anki.api.AddContentApi

// Thin Kotlin wrapper over AnkiDroid's [AddContentApi] (FR-1.3.3, T4.1). Resolves
// decks / models by name (create-or-find, since the API exposes no by-name
// lookup), reports install state, and fronts the explicit first-field dupe
// check the export service needs before adding. Every call hits AnkiDroid's
// content provider, so callers run this off the platform main thread.
//
// AnkiDroid's addNote performs NO duplicate checking — it always inserts.
// Re-save safety comes from [noteExists] (Anki keys uniqueness on a model's
// first field).
//
// T16.1: install detection + the runtime-permission gate live here so the
// channel (T16.2) + Dart UI (T16.3) drive AnkiDroid's native grant flow before
// the first content-provider query — without it every deckList/addNote call is
// rejected with "Permission not granted for: CardContentProvider.query".
class AnkiGateway(private val context: Context) {

    private val api: AddContentApi by lazy { AddContentApi(context) }

    /**
     * Whether AnkiDroid is installed AND its API is exposed (T16.1). Uses
     * [AddContentApi.getAnkiDroidPackageName] — the canonical check from the
     * official apisample — which returns null when AnkiDroid is absent or when
     * its global "Enable AnkiDroid API" toggle is off (AnkiDroid 2.24+).
     * Replaces the raw packageManager probe, which reported "installed" even
     * when the API was disabled.
     */
    fun isInstalled(): Boolean = AddContentApi.getAnkiDroidPackageName(context) != null

    /**
     * Whether the AnkiDroid READ_WRITE permission still needs a runtime grant
     * (T16.1). The v1.1.0 aar ships no `shouldRequestPermission` on
     * [AddContentApi], so this reimplements the apisample's
     * `AnkiDroidHelper.shouldRequestPermission` inline: post-M material release
     * + [ContextCompat.checkSelfPermission] against the aar's exported
     * [AddContentApi.READ_WRITE_PERMISSION] constant. Mirrors what
     * `AddContentApi.hasReadWritePermission()` (private) checks internally.
     */
    fun shouldRequestPermission(): Boolean =
        Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
            ContextCompat.checkSelfPermission(
                context,
                AddContentApi.READ_WRITE_PERMISSION,
            ) != android.content.pm.PackageManager.PERMISSION_GRANTED

    /** Create-or-find the named deck. Returns its id. */
    fun ensureDeck(name: String): Long {
        api.deckList?.entries?.firstOrNull { it.value == name }?.let { return it.key }
        return api.addNewDeck(name)
            ?: throw IllegalStateException("AnkiDroid returned no deck id for \"$name\"")
    }

    /** Create-or-find a single-card note type by name. Returns its model id. */
    fun ensureModel(
        name: String,
        fields: List<String>,
        front: String,
        back: String,
        css: String,
    ): Long {
        api.modelList?.entries?.firstOrNull { it.value == name }?.let { return it.key }
        val mid = api.addNewCustomModel(
            name,
            fields.toTypedArray(),
            arrayOf("Card 1"),
            arrayOf(front),
            arrayOf(back),
            css.ifEmpty { null },
            null,
            null,
        ) ?: throw IllegalStateException("AnkiDroid returned no model id for \"$name\"")
        return mid
    }

    /** Whether a note whose first field equals [firstField] exists for [modelId]. */
    fun noteExists(modelId: Long, firstField: String): Boolean =
        api.findDuplicateNotes(modelId, firstField).isNotEmpty()

    /** Always inserts (no dupe check). Returns the note id, or null on failure. */
    fun addNote(
        deckId: Long,
        modelId: Long,
        fields: List<String>,
        tags: List<String>,
    ): Long? = api.addNote(modelId, deckId, fields.toTypedArray(), tags.toSet())

    /**
     * Import an image into AnkiDroid's media collection (T4.4, FR-1.3.4).
     * [fileUri] must be readable by AnkiDroid — the caller exposes a FileProvider
     * content URI and grants read permission to com.ichi2.anki. [preferredName]
     * is the base filename (no extension) AnkiDroid stores the media under.
     * Returns the formatted `<img src="...">` field string to drop into a note,
     * or null if AnkiDroid could not import the media (retryable — the cached
     * image is unaffected, so a later export re-tries).
     */
    fun addMedia(fileUri: Uri, preferredName: String): String? =
        api.addMediaFromUri(fileUri, preferredName, "image")
}

File: app/lib/features/anki/application/anki_gateway.dart
// Abstract seam over AnkiDroid (M4, FR-1.3.3, T4.1). The platform impl talks
// to AnkiDroid via its API; tests swap in a fake. Keeping the seam local —
// names, not ids — lets the caller stay in the domain while the adapter
// resolves ids and hides AnkiDroid's content-provider details.
//
// Idempotency: AnkiDroid's addNote performs NO duplicate checking — it always
// inserts. Callers must guard re-saves by checking [hasNoteWithFirstField]
// first (Anki keys uniqueness on the model's first field). This is the
// re-save-safety the spec asks for (FR-1.3.3).

import 'package:rivendell/features/anki/domain/anki_model_spec.dart';

abstract class AnkiGateway {
  /// Whether AnkiDroid is installed and its API is reachable. False triggers
  /// the "install AnkiDroid" affordance in the UI.
  Future<bool> isInstalled();

  /// Whether the AnkiDroid READ_WRITE permission still needs a runtime grant
  /// (T16.1). True → the UI shows a one-time explainer, then calls
  /// [requestPermission]. Read live from the API, never cached — repeated
  /// exports must not re-prompt once granted.
  Future<bool> shouldRequestPermission();

  /// Drive AnkiDroid's native runtime-permission grant screen (T16.2).
  /// Returns true if the grant succeeded. On AnkiDroid 2.24+ the user must
  /// also have the global "Enable AnkiDroid API" toggle ON, else this returns
  /// false and the UI surfaces current-copy guidance.
  Future<bool> requestPermission();

  /// Create the deck [name] if absent, else return its id. Stable across calls.
  Future<int> ensureDeck(String name);

  /// Create the note type [spec] if absent, else return its id. Stable across
  /// calls; looked up by [AnkiModelSpec.name].
  Future<int> ensureModel(AnkiModelSpec spec);

  /// Whether a note whose first field equals [firstField] already exists for
  /// [modelId]. The idempotency guard the export service checks before adding.
  Future<bool> hasNoteWithFirstField({
    required int modelId,
    required String firstField,
  });

  /// Add a note to [deckId] using [modelId]. [fields] must match the model's
  /// field order; [tags] label the note (e.g. the recording filename). Always
  /// inserts — callers must dedupe via [hasNoteWithFirstField].
  ///
  /// Returns the new note id, or null if the insert failed (not on dupe).
  Future<int?> addNote({
    required int deckId,
    required int modelId,
    required List<String> fields,
    required Set<String> tags,
  });

  /// Add an image to AnkiDroid's media collection and return the formatted
  /// field string to drop into a note (`<img src="...">`), or null on failure.
  ///
  /// [relativePath] is app-relative under the documents dir (where the AI image
  /// cache writes, e.g. `ai_images/<sha1>.png`); the platform side resolves it
  /// against `filesDir` and exposes it to AnkiDroid via a FileProvider content
  /// URI with a read grant. [preferredName] is the base filename (no
  /// extension) AnkiDroid stores the media under. Returns null when AnkiDroid
  /// can't import the media — callers treat that as a retryable failure (the
  /// image is still cached, so a later export re-tries the attach).
  Future<String?> addMedia({
    required String relativePath,
    required String preferredName,
  });
}
